bees/emailserverbee/emailserverbee.go

```diff
@@ -0,0 +1,193 @@
+/*
+ *    Copyright (C) 2014-2017 Christian Muehlhaeuser
+ *
+ *    This program is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU Affero General Public License as published
+ *    by the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *    Authors:
+ *      Matthias Krauser <matthias@krauser.eu>
+ */
+
+package emailserverbee
+
+import (
+	"github.com/flashmob/go-guerrilla"
+	"github.com/flashmob/go-guerrilla/backends"
+	"github.com/flashmob/go-guerrilla/mail"
+
+	//"github.com/flashmob/go-guerrilla/response"
+
+	"github.com/muesli/beehive/bees"
+)
+
+// EmailServerBee is a Bee that starts an SMTP server and fires events for incoming
+// Emails.
+type EmailServerBee struct {
+	bees.Bee
+
+	address        string
+	allowedHosts   []string
+	hostname       string
+	maxSize        int64
+	timeout        int
+	maxClients     int
+	startTLSOn     bool
+	tlsAlwaysOn    bool
+	privateKeyFile string
+	publicKeyFile  string
+
+	eventChan chan bees.Event
+}
+
+func (mod *EmailServerBee) mailProcessor() func() backends.Decorator {
+	return func() backends.Decorator {
+		return func(p backends.Processor) backends.Processor {
+			return backends.ProcessWith(
+				func(e *mail.Envelope, task backends.SelectTask) (backends.Result, error) {
+					if task == backends.TaskSaveMail {
+
+						recipients := make([]string, len(e.RcptTo))
+
+						for index, rcpt := range e.RcptTo {
+							recipients[index] = rcpt.String()
+						}
+
+						/*headers := make([]string, len(e.Header))
+
+						i := 0
+						for name, value := range e.Header {
+							headers[i] = name + ": " + value
+							i += 1
+						}*/
+
+						// create events and send it to cin
+						ev := bees.Event{
+							Bee: mod.Name(),
+							Options: []bees.Placeholder{
+								{
+									Name:  "sender",
+									Type:  "string",
+									Value: e.MailFrom.String(),
+								},
+								{
+									Name:  "remote_ip",
+									Type:  "string",
+									Value: e.RemoteIP,
+								},
+								{
+									Name:  "recipients",
+									Type:  "string[]",
```

Comment:
This should read []string, as we decided to stick with Go's conventions. It currently hasn't much of an effect within Beehive, but the admin interface will make use of that information, soon.